src/models/marian.cpp

```diff
@@ -0,0 +1,244 @@
+    return;
+  }
+
+  // Store length of input sequence for each batch for the get step
+  for (int b = 0; b < state_.params_->search.batch_size; b++) {
+    // Find the first non pad token from the end
+    size_t token_index = new_kv_length;
+    while (token_index-- > 0) {
+      auto next_token = const_cast<DeviceSpan<int32_t>&>(next_tokens).CpuSpan()[b * new_kv_length + token_index];
+      if (next_token != model_.config_->model.pad_token_id)
+        break;
+    }
+    input_sequence_lengths[b] = static_cast<int>(token_index + 1);
+  }
+
+  if (output_raw_->ort_tensor_ && static_cast<size_t>(output_raw_->GetShape()[1]) == new_kv_length) {
+    return;
+  }
+
+  output_raw_->CreateTensor(shape_, state_.params_->use_graph_capture);
+  state_.outputs_[output_index_] = output_raw_->GetOrtTensor();
+}
+
+void MarianLogits::Add() {
+  output_index_ = state_.outputs_.size();
+
+  state_.output_names_.push_back(model_.config_->model.decoder.outputs.logits.c_str());
+  state_.outputs_.push_back(output_raw_->GetOrtTensor());
+}
+
+MarianInputIDs::MarianInputIDs(State& state)
+    : state_{state} {
+  name_ = model_.config_->model.decoder.inputs.input_ids.c_str();
+  shape_ = {state_.params_->BatchBeamSize()};
+  type_ = model_.session_info_.GetInputDataType(name_);
+
+  value_ = std::make_unique<Tensor>(model_.p_device_inputs_, Ort::TypeToTensorType<int32_t>);
+  cast_value_ = std::make_unique<Tensor>(model_.p_device_inputs_, Ort::TypeToTensorType<int64_t>);
+}
+
+void MarianInputIDs::AddMarianInputs() {
+  input_index_ = state_.inputs_.size();
+
+  state_.inputs_.push_back(value_->GetOrtTensor());
+  state_.input_names_.push_back(name_);
+}
+
+void MarianInputIDs::Update(DeviceSpan<int32_t> new_tokens) {
+  auto new_tokens_cpu = new_tokens.CopyDeviceToCpu();
+
+  const auto get_unpadded_sequence_length = [](std::span<const int32_t> input_ids, int32_t pad_token_id) {
+    for (int32_t i = 0; i < input_ids.size(); i++) {
+      if (input_ids[i] == pad_token_id)
+        return i;
+    }
+    return static_cast<int32_t>(input_ids.size());
+  };
+
+  value_->CreateTensor(shape_, state_.params_->use_graph_capture);
+  state_.inputs_[input_index_] = value_->GetOrtTensor();
+
+  // Update input_ids with next tokens
+  auto data_span = value_->GetDeviceSpan<int32_t>();
+  data_span.CopyFrom(new_tokens);
+
+  if (type_ == Ort::TypeToTensorType<int64_t>) {
+    if (!cast_value_->ort_tensor_)
+      cast_value_->CreateTensor(shape_, state_.params_->use_graph_capture);
+    Cast(*value_->GetOrtTensor(), cast_value_->ort_tensor_, *model_.p_device_inputs_, type_);
+    state_.inputs_[input_index_] = cast_value_->GetOrtTensor();
+  }
+
+  is_prompt_ = false;
+}
+
+DeviceSpan<float> MarianState::Run(int current_length, DeviceSpan<int32_t>& next_tokens, DeviceSpan<int32_t> next_indices) {
+  if (first_run_) {
+    // INITIALIZE THE ENCODER AND RUN IT ONCE
```

Comment:
nit: avoid capitalization unless it is a warning comment.

```suggestion
    // Initialize the encoder and run it once
```